### DISCUSSION TITLE
Replace editor change handling with command pattern

### DISCUSSION THREAD

**Discussion by OliBomby** [ppy/osu#25959] [Ideas]
So change handling in the osu! editor is a bit weird. It relies on the legacy encoding to save states and line diff for undoing. Also the editor beatmap has to manually be notified about any changes to hit objects, so it can update stacking and combo information.

The command design pattern provides a better alternative. It can store state changes as a stack of commands which each represent an atomic change in the beatmap state, and the commands can replace the need to explicitly notify the editor about state changes to hit objects. It will be more future proof because there is no need for the legacy encoder, and it makes potential future editor features like live-share more realistic.

The arguments for the command design pattern also apply to the skin editor.

Let me know what you think and if this is something I should implement for the osu! editor.

# Design

Here is a design I came up with for some relevant classes and interfaces that would be involved in the beatmap editor if it used the command pattern.

`ICommandHandler`
Interface for anything that accepts commands.
- `Apply(Command command)`

`CommandHandler : TransactionalCommitComponent`
Responsible for logging commands and handling undo logic. Optional component in editor.
- `BeginCommit(string name)` adds a special marker on the undo stack that indicates this is the start of the commit. This ensures undo undoes all the commands since the user action. Also any commands that were applied outside of a commit (like beatmap processing) will be undone until the user action.
- `Apply(Command command)` executes the command if not done already and adds it to the undo stack, then clears the redo stack.
- `Undo()` undoes commands on the stack until a special marker is reached. Each undone command gets added to the redo stack.
- `Redo()` redoes commands until the second special market is reached (not inclusive). Each redone command gets added to the undo stack.

`CommitMarkerCommand : Command`
Special command which does nothing which indicates a start of commit in the undo stack.
- `string Name`

`EditorBeatmap : ICommandHandler`
Responsible for broadcasting changes. Receives all commands and forwards them to `CommandHandler` if that exists.

`Command`
Base class for commands.
Would be nice if this class is serializable, but that is probably a pain in the ass to implement.
- `Apply(ICommandHandler commandHandler)` applies the command to the given command handler. Only executes if `commandHandler` is not null. Allows syntax like `new MoveCommand(hitObject, pos).Apply(EditorBeatmap)`. This could be done through an extension method instead of inheritance.
- `Do()` executes the command. Sets `IsExecuted` to true.
- `Undo()` executes the opposite of the command. Sets `IsExecuted` to false.
- `bool IsExecuted` prevents double execution. `Apply()` can use this to make sure the command is always executed once.

`IHitObjectUpdateCommand`
Interface for commands that update a property of a `HitObject`. Lets the `EditorBeatmap` detect which hit objects are being updated by the applied commands.
- `HitObject Target`
Also have these for adds, removes, and timing points.

`IMergeableCommand`
Interface used for merging multiple subsequent commands into one. This will significantly reduce the undo stack size when dragging objects.
The `CommandHandler` will try to merge the current command with the previous command and if so, remove the previous command from history.
- `bool TryMerge(Command command)` tells whether this command overrides the effect of a previous command. If so, merges the history of that command into this one.

`MultiMoveCommand : Commmand, IMergeableCommand, IMultiHitObjectUpdateCommand`
Command that moves multiple hit objects at once along a single direction. 
Instead of making a bunch of individual move commands which can't be easily merged, using one big move command is a lot more efficient.
- `OsuHitObject[] Targets`
- `Vector2 StartPosition`
- `Vector2 EndPosition`
Make sure multiple moves and merges dont make the history innacurate due to float precision.

**Comment by bdach** [comment]
Since this is on my mind because of [another comment elsewhere](https://github.com/ppy/osu/discussions/35392#discussioncomment-14888236) I want to dump a particular piece of information off my brain about this replacement that I don't think I've ever vocalised anywhere public.

I tried this myself a few months ago and the reason why this is pretty difficult to do correctly and pretty much impossible to do gradually is because the current way which is based on the `.osu` format is a bit too good with respect to one aspect of change tracking, which is *object identity*.

The snapshot strategy of undo, which we're using, allows us not to care about the identity of any particular object that's being mutated. We don't need to keep track whether the hitcircle being restored after a delete is the same hitcircle that's been moved seventeen hundred times in fifty five undo steps preceding it; we just have a state dump that we restore period.

However when moving to change tracking via commands / deltas this becomes a *very* painful issue very fast. To undo an operation you want to be able to apply an undo to a *specific* object that probably has a historical lifetime, and now you need to be able to *track* that lifetime. Otherwise you're going to undo an object deletion, which will spawn *an entire different* object, and all of the undo steps that touched that deleted object will actually fail to apply because the change handling system doesn't understand that they're the same object.

To that end, you either need to either
- hard rely on reference equality of the objects, which makes it essentially impossible to reconcile any new change handler with the current snapshot-based one,
- introduce some sort of object identification measure like a guid on every hitobject / thing you're mutating, which could maybe work with the old change handler, *as long as* you serialise the guid out to the beatmap somehow, and probably in a way that doesn't just completely obliterate stable compatibility. Maybe given we have `.olz` now as an escape, and the legacy export facility to boot, we can abide by this...?

**Reply by OliBomby** [reply]
There is another reason why it is essentially impossible to reconcile any new change handler with the current snapshot-based one: You can not turn off tracking for certain types of changes in the snapshot-based change handler. If you want to have hit object movement handled by the command-based change handler and everything else by the old change handler, the old change handler is still going to track and make snapshots for any hit object movements. This makes it very hard to determine which change handler you should use for any undo operation, and it just adds a lot of complexity which you don't want to deal with.

My solution in #30314 just hard relies on reference equality and sidesteps reconciling with the current snapshot-based change handler by porting all operations on hit objects to the new change handler. This makes it so the old and new change handlers work on entirely disjoint object sets, so there is no issues with the old change handler spawning entirely different hit objects. I think this is the simplest solution possible.

Adding unique identifiers has great benefits like no longer relying on reference equality, being able to persist deltas locally, or transfer deltas over the network. This should be possible to add afterwards when we already have a simple pointer reference based implementation.

**Reply by bdach** [reply]
> You can not turn off tracking for certain types of changes in the snapshot-based change handler

I'm not so sure about this one. Maybe I haven't gone far enough to see why it doesn't work but my angle was basically something like https://github.com/ppy/osu/commit/6cd3c037f64d932ade3775e94f20f86707f85837 ([full branch](https://github.com/ppy/osu/compare/master...bdach:osu:change-handling-rnd)) - the angle is to encapsulate the old application of snapshots into a "command" of its own, and have all existing code paths go through that to emit "snapshot commands", and then successively replace usage of the old code paths with submitting new-style commands.

**Reply by OliBomby** [reply]
> > You can not turn off tracking for certain types of changes in the snapshot-based change handler
> 
> I'm not so sure about this one. Maybe I haven't gone far enough to see why it doesn't work but my angle was basically something like [6cd3c03](https://github.com/ppy/osu/commit/6cd3c037f64d932ade3775e94f20f86707f85837) ([full branch](https://github.com/ppy/osu/compare/master...bdach:osu:change-handling-rnd)) - the angle is to encapsulate the old application of snapshots into a "command" of its own, and have all existing code paths go through that to emit "snapshot commands", and then successively replace usage of the old code paths with submitting new-style commands.

Okay maybe I'm misremembering things, because this particular issue doesn't seem that relevant in your implementation. As long as you update the `lastSnapshot` but dont add a snapshot command to history any time a command is submitted, it should be separated fine.

Then I agree with the two options layed out. Either ensure reference equality and port all everything at once, or introduce guids to objects.

Guids come with their own set of challenges like:
- How to serialize it without breaking stable or external tools?
- How to reference sub-parts of hit objects, like individual slider anchors or node samples? Do we give these their own guids too or refer to them by index?
- Can we refer to everything with paths instead of guids? Like for example `hitobjects[10].path.control_points[3]`. Maybe this prevents having to change the serialization?

**Reply by minetoblend** [reply]
Since I've been using the Guid approach in my own undo/redo system(s) for quite a while now I thought I'd chime in on this a little. Quick disclaimer: I only have experience with approaches that support sending deltas over network (guids & paths as described above) and have never attempted to build one that rely on references for tracking object identities.

> - How to serialize it without breaking stable or external tools?

Do object identities need to be persisted with the beatmap at all? The undo/redo stacks will get discarded when you close the editor so unless those ids will be used in other places too it should be fine to just generate them from scratch whenever you open a beatmap in the editor.
I can maybe see the need for it if you try to gradually transition to the new changehandler while keeping the old one around, but even if you do this it can be avoided. If both change handlers have to co-exist for a while I see 3 ways of doing this (in ascending order of footgun count):

- gradually introduce commands to perform mutations on the beatmap, but keep the old change handler in charge of undo/redo. The commands would simply apply the change & the rest keeps working through the old change handler before. Once everything has been converted to use commands, the underlying change handler can be swapped out for the new one which makes use of the commands for undo/rdeo.

- Do use both change handlers for undo/redo at the same time but transition anything that *doesn't* rely on object identity or is incompatible with the old change handler in other ways to the new changehandler first. Every change that *is* incompatible would have to get lined up as a series of prs to get all merged in close succession so the old change handler can be made obsolete within a release cycle.

- Let both change handlers have a free-for-all and try to transition hitobjects or timing points to the new change handler first.

> - How to reference sub-parts of hit objects, like individual slider anchors or node samples? Do we give these their own guids too or refer to them by index?

Is there any benefit from storing changes with that much granularity? I say just store the whole thing, that approach has been serving me quite well so far and the lack of granularity has never been an issue. If the memory overhead ever becomes a problem here (which shouldn't happen as long as you limit the history to a reasonable length), deltas can be stored on disk as long as they're fully serializable (with guids you're already halfway there for that).

> - Can we refer to everything with paths instead of guids? Like for example hitobjects[10].path.control_points[3]. Maybe this prevents having to change the serialization?

That's what I did before I switched to just using ids, I found this approach to be quite annoying because it introduces complexity to solve problems that shouldn't even exist. You're gonna need a mechanism to both create & resolve these paths, which is likely gonna need maintenance down the road. I think simply reducing the granularity to the point where you don't need it is a far simpler solution. If there ever are any commands where there is truly no way around storing object identities in a more granular way, then that logic should be self-contained within those specific commands' classes and not creep its way into other commands.

---

In practice the above listed concerns have been non-issues when I was working with guids, most of these problems can be avoided by just keeping things simple. Overall I see Guids as the approach with the least amount of footguns.
